File: c++/56. Merge Intervals.cpp
#include <iostream>
#include <vector>

using namespace std;

struct Interval {
    int start;
    int end;
    Interval() : start(0), end(0) {}
    Interval(int s, int e) : start(s), end(e) {}
};
 
 
 bool cmp(Interval a, Interval b){
     if(a.start != b.start)
        return a.start < b.start;
    return a.end > b.end;
 }
 
class Solution {
public:
    vector<Interval> merge(vector<Interval>& intervals) {
        sort(intervals.begin(), intervals.end(), cmp);
        vector<Interval> merged;
        Interval last;
        if(intervals.size() == 0)
            return merged;
        last = intervals[0];
        for(int i=1;i<intervals.size();i++)
            if(intervals[i].start <= last.end)
                last.end = max(last.end, intervals[i].end);
            else{
                merged.push_back(last);
                last = intervals[i];
            }
        merged.push_back(last);
        return merged;
    }
};

int main(){
    vector<Interval> v;
    for(int i=0;i<10;i++)
        v.push_back(Interval(i, i+2));
    Solution s;
    v = s.merge(v);
    for(int i=0;i<v.size();i++)
        cout << v[i].start << ' ' << v[i].end << endl;
}

File: java/56. Merge Intervals.java
class Solution {

    private static final int MAX_NUMBER = 10000;

    public int[][] merge(int[][] intervals) {
        ArrayList<Interval> resultIntervals = new ArrayList<Interval>();

        int[] intervalMarkArray = new int[10005];

        // Mark the start and end points.
        for (int i = 0; i < intervals.length; i++) {
            // Assert it has 2 elements as expected.
            int start = intervals[i][0];
            int end = intervals[i][1];
            intervalMarkArray[start]++;
            intervalMarkArray[end]--;
        }

        int currentStart = 0;
        for (int i = 1; i < MAX_NUMBER; i++) {
            intervalMarkArray[i] = intervalMarkArray[i] + intervalMarkArray[i - 1];
            if (intervalMarkArray[i] == 1 && intervalMarkArray[i - 1] == 0) {
                currentStart = i;
            } else if (intervalMarkArray[i] == 0 && intervalMarkArray[i - 1] == 1) {
                resultIntervals.add(new Interval(currentStart, i));
                currentStart = i + 1;
            }
        }

        // Convert interval back to array.
        int[][] resultArray = new int[resultIntervals.size()][2];
        for (int i = 0; i < resultIntervals.size(); i = i + 1) {
            resultArray[i][0] = resultIntervals.get(i).start;
            resultArray[i][1] = resultIntervals.get(i).end;
        }

        return resultArray;
    }

    class Interval {
        int start;
        int end;

        Interval(int start, int end) {
            this.start = start;
            this.end = end;
        }
    }
}